open_lm/data.py

```diff
@@ -56,6 +57,39 @@ def preprocess_json(text, vocab_size):
     return text
 
 
+def _batched_fulldata(data, batchsize=20, collation_fn=wds.filters.default_collation_fn, partial=True):
+    batch = []
+    first_batch = None
+    for sample in data:
+        if len(batch) >= batchsize:
+            if first_batch == None:
+                first_batch = batch
+            if collation_fn is not None:
+                batch = collation_fn(batch)
+            yield batch
+
+            batch = []
+        batch.append(sample)
+
+    if len(batch) == 0:
+        return
+    elif len(batch) == batchsize or partial:
+        if collation_fn is not None:
+            batch = collation_fn(batch)
+        yield batch
+    elif len(batch) < batchsize and not partial:
+        for sample in first_batch:
```

Comment:
This was the intended behavior of batched_fulldata (so that the last few samples are not dropped). Since the number of batches is selected in a way that the last few samples aren't seen, the behavior ends up being essentially the same as wds.batched (and dropping the last batch if partial is False is exactly what wds.batched does, and is its only difference with batched_fulldata).

This function is mostly a leftover from a previous iteration, so we can also just remove it and go back to using wds.batched.